librz/util/axml.c

```diff
@@ -0,0 +1,445 @@
+			return NULL;
+		}
+
+		// UTF-8 encoded length
+		n = *start++;
+		if (n & 0x80) {
+			n = ((n & 0x7f) << 8) | *start++;
+		}
+
+		name = calloc(n + 1, 1);
+
+		if (n == 0) {
+			if (length) {
+				*length = 0;
+			}
+
+			return name;
+		}
+
+		if ((ut64)start > (ut64)data + data_size - sizeof(ut32) - n - 1) {
+			free(name);
+			return NULL;
+		}
+
+		memcpy(name, start, n);
+
+		if (length) {
+			*length = n;
+		}
+	} else {
+		if ((ut64)start > (ut64)data + data_size - sizeof(ut32)) {
+			return NULL;
+		}
+
+		ut16 *start16 = (ut16 *)start;
+
+		// If >0x7fff, stored as a big-endian ut32
+		ut32 n = rz_read_le16(start16++);
+		if (n & 0x8000) {
+			n |= ((n & 0x7fff) << 16) | rz_read_le16(start16++);
+		}
+
+		// Size of UTF-16LE without NULL
+		n *= 2;
+
+		name = calloc(n * 2 + 1, 1);
+
+		if ((ut64)start16 > (ut64)data + data_size - sizeof(ut32) - n - 1) {
+			free(name);
+			return NULL;
+		}
+
+		// If UTF-16LE, decode to UTF-8 so we can print it to the screen
+		if (rz_str_utf16_to_utf8((ut8 *)name, n * 2, (const ut8 *)start16, n, true) < 0) {
+			free(name);
+			eprintf("Failed to decode UTF16-LE\n");
+			return NULL;
+		}
+
+		if (length) {
+			*length = n;
+		}
+	}
+
+	return name;
+}
+
+static char *resource_value(string_pool_t *pool, const ut8 *data, ut64 data_size,
+	resource_value_t *value) {
+	switch (value->type) {
+	case RESOURCE_NULL:
+		return rz_str_new("");
+	case RESOURCE_REFERENCE:
+		return rz_str_newf("@0x%x", value->data.d);
+	case RESOURCE_STRING:
+		return string_lookup(pool, data, data_size, rz_read_le32(&value->data.d), NULL);
+	case RESOURCE_FLOAT:
+		return rz_str_newf("%f", value->data.f);
+	case RESOURCE_INT_DEC:
+		return rz_str_newf("%d", value->data.d);
+	case RESOURCE_INT_HEX:
+		return rz_str_newf("0x%x", value->data.d);
+	case RESOURCE_BOOL:
+		return rz_str_newf(value->data.d ? "true" : "false");
+	default:
+		eprintf("Resource type is not recognized: %#x\n", value->type);
+		break;
+	}
+	return rz_str_new("null");
+}
+
+static bool dump_element(RzStrBuf *sb, string_pool_t *pool, namespace_t *namespace,
+	const ut8 *data, ut64 data_size, start_element_t *element,
+	const ut32 *resource_map, ut32 resource_map_length, st32 *depth, bool start) {
+	ut32 i;
+
+	char *name = string_lookup(pool, data, data_size, rz_read_le32(&element->name), NULL);
+	for (i = 0; i < *depth; i++) {
+		rz_strbuf_appendf(sb, "\t");
+	}
+
+	if (start) {
+		rz_strbuf_appendf(sb, "<%s", name);
+		ut16 count = rz_read_le16(&element->attribute_count);
+		if (*depth == 0 && namespace) {
+			char *key = string_lookup(pool, data, data_size, namespace->prefix, NULL);
+			char *value = string_lookup(pool, data, data_size, namespace->uri, NULL);
+			rz_strbuf_appendf(sb, " xmlns:%s=\"%s\"", key, value);
+			free(key);
+			free(value);
+		}
+
+		if (count != 0) {
+			rz_strbuf_appendf(sb, " ");
+		}
+
+		for (i = 0; i < count; i++) {
+			attribute_t a = element->attributes[i];
```

Comment:
I suggest extracting what's inside the loop into a separate function.

Reply:
minor thing TBH. This is a backport, not a rewrite.